### PR TITLE
queue data structure

### PR DESCRIPTION
**Pull request by Compsight**

### CODE REVIEW

File: arrayDraftFormat [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+  context('push()', () => {
+    it('pushes an element to the top of the stack.', () => {
+      const myStack = new Stack()
+
+      expect(() => myStack.push('foo'))
+        .to.alter(() => myStack.length(), { from: 0, to: 1 })
+    })
+  })
+})
```
Comment (nicosesma) [inline arrayDraftFormat:48]: is this duplicate code?

File: arrayDraftFormat [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
```
Comment (nicosesma) [inline arrayDraftFormat:166]: what is this all supposed to be?

File: spec/queueSpec.js [hunk 3]
```diff
@@ -0,0 +1,83 @@
+import chai, { expect } from 'chai'
+import chaiChange from 'chai-change'
+import Queue from '../src/queue'
+
+chai.use(chaiChange)
+
+describe.only('Queue', () => {
```
Comment (nicosesma) [inline spec/queueSpec.js:7]: it would be better to remove the ```describe.only```  if this is going to be merged

File: spec/queueSpec.js [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    })
+  })
+
+  context("dequeue()", () => {
+    it("returns and removes the front element in the queue", () => {
+      const testDequeue = new Queue()
+      testDequeue.enqueue("foo")
+
+      expect(testDequeue.dequeue()).to.equal("foo")
+    })
```
Comment (nicosesma) [inline spec/queueSpec.js:36]: would you be able to add a test to show that the element was removed?
Such as,
```
const set = new Queue()
set.enqueue('thing')
expect( () => set.dequeue()).to.alter( () => set.length(), { from: 1, to: 0 })
```

File: spec/queueSpec.js [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+  context("enqueue(\"foo\")", () => {
+    it("adds an element to the back of the queue", () => {
+      const testEnQueue = new Queue()
+      // testEnQueue.enqueue("foo")
+
+      expect(() => testEnQueue.enqueue("foo"))
+        .to.alter(() => testEnQueue.length(), {from:0, to:1})
+      expect(() => testEnQueue.enqueue("bar"))
+        .to.alter(() => testEnQueue.back(), {from:"foo", to:"bar"})
```
Comment (nicosesma) [inline spec/queueSpec.js:53]: It's great that you're testing this

File: src/queue.js [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    return this.count === 0
+      ? null
+      : this.begin.data
+  }
+
+  isEmpty() {
+    return this.count === 0
+      ? true
+      : false
+  }
```
Comment (nicosesma) [inline src/queue.js:52]: this could even be
```
return (this.count === 0)
```

File: src/queue.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
+'use strict'
+
+class Node {
+  constructor(data, follows, proceeds) {
+    this.follows = follows
+    this.data = data
+    this.proceeds = proceeds
+  }
+}
+
```
Comment (nicosesma) [inline src/queue.js:10]: The use of the Node class is a brilliant manner of separating the responsibility of the functions

File: src/queue.js [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+  dequeue() {
+    const next = this.begin
+    return this.count === 0
+      ? next
+      :(this.begin = next.proceeds,
+        this.count--,
+        next.data)
+  }
+
```
Comment (nicosesma) [inline src/queue.js:32]: The use of these ternary operators is great

File: src/queue.js (hunk 7 above)
Comment (sashajustice) [reply src/queue.js:10]: I think I like your comment but what do you mean?

**Pull request merged**